A80-pid.md

```diff
@@ -0,0 +1,290 @@
+  }
+
+  // use value calculated in the onEDFSchedulerUpdate method
+  meanUtilization = data.meanUtilization
+
+  // call the PID controller to get the value of the control signal.
+  controlSignal = data.pidController.update({
+    referenceSignal:  meanUtilization,
+    actualSignal:     utilization,
+    samplingInterval: time.Since(lastApplied),
+  })
+
+  // Normalize the signal.
+  // If meanUtilization ~= 0 the signal will be ~= 0 as well, and convergence will become painfully slow.
+  // If, meanUtilization >> 1 the signal may become very high, which could lead to oscillations.
+  if meanUtilization > 0 {
+    controlSignal *= 1 / meanUtilization
+  }
+
+  lastAppliedWeight = data.lastAppliedWeightPerSubchannel[subchannelID]
+
+  // Use controlSignal to adjust the weight.
+  // First calculate a multiplier that will be used to determine how much weight should be changed.
+  // The higher is the absolute value of the controlSignal the more we need to adjust the weight.
+  if controlSignal >= 0 {
+    // in this case mult should belong to the [1,inf) interval, so we will be increasing the weight.
+    mult = 1.0 + controlSignal
+  } else {
+    // in this case mult should belong to (0, 1) interval, so we will be decreasing the weight.
+    mult = -1.0 / (controlSignal - 1.0)
+  }
+  weight = lastAppliedWeight * mult
+
+  // Clamp weight
+  if weight > conf.MaxWeight {
+    weight = conf.MaxWeight
+  }
+  if weight < conf.MinWeight {
+    weight = conf.MinWeight
+  }
+
+  // Save resulting utilization and weight.
+  data.utilizationPerSubchannel[subchannelId] = utilization
+  data.lastAppliedWeightPerSubchannel[subchannelID] = weight
+
+  return weight
+}
+
+func onEDFSchedulerUpdate(data callbackData) {
+  // Calculate mean utilization across all subchannels
+  totalUtilization = 0
+  count = len(data.utilizationPerSubchannel)
+  for _, utilization in data.utilizationPerSubchannel {
+    totalUtilization += utilization
+  }
+  data.meanUtilization = totalUtilization / count
+}
+```
+
+The proposal is to make `wrrCallbacks` public. This has a number of significant benefits. Besides PID, there are other cases where one might need to extend `wrr`. For example, Spotify [demonstrates](https://www.youtube.com/watch?v=8E5zVdEfwi0) a gRPC load balancer to reduce cross-zone traffic – this can be implemented nicely in terms of `wrr` weights. We are also considering the same and incorporating things like latency into our load balancing decisions. Existing ORCA extension points don't cover these use cases. We leverage ORCA for custom server utilization metrics, but we also need the ability to combine server and client metrics to generate the resulting weight. The alternative is to write our own balancer with custom EDF scheduler and handle details related to subchannel management and interactions with resolvers. With this new API, use cases like this can be covered naturally, users have full control over the end-to-end definition of weights.
```

Comment:
I am hesitant to make this a public API.  This seems like a valuable internal thing to allow us to reuse LB policy implementations, but making it public implies a stability commitment that I'm not super comfortable with.

It's worth noting that the example of using latency would require that this API surface be even broader, because we'd need a way for this to trigger the LB policy to measure the latency on each RPC.  (We do have that mechanism internally, but we'd need to expose it via this API in addition to the existing mechanism as part of the LB policy API itself.)

I'd like to hear thoughts from @ejona86 and @dfawley on this.

Reply:
I expected this answer but decided to give it a try anyway. Making this interface private works for us as well - next time we need to extend PID of WRR we'll come back with more gRFCs. I assume that it is much easier to make an interface public than otherwise, so maybe we can revisit this decision in the future when we have more data and better defined use-cases where such interface might be useful. 

For now I can remove this paragraph and replace it with a note that the interface will be private. @markdroth does this work for you?

Reply:
I'm not sure that I'd want exactly this interface in C-core, even as a private API.  I think it would be better to avoid being proscriptive about how the implementation needs to be structured, so that each implementation can make its own choice.

That having been said, I do agree that it's useful to point out that there's so much overlap between this new policy and the existing WRR policy and to suggest that implementations consider how to avoid code duplication.  And we can even describe this specific approach as one possible way to do that.  Let's just leave the final decision up to the implementation.

FWIW, I'm going to be putting together a gRFC soon to allow WRR to configure which ORCA utilization metrics it looks at, to make it a bit more flexible.  I don't think that actually affects this design, though, because it still doesn't make the weight calculation function pluggable.